sdks/python/apache_beam/runners/interactive/pipeline_instrument_test.py

```diff
@@ -178,14 +183,48 @@ def test_has_unbounded_source(self):
     p = beam.Pipeline(interactive_runner.InteractiveRunner())
     _ = p | 'ReadUnboundedSource' >> beam.io.ReadFromPubSub(
         subscription='projects/fake-project/subscriptions/fake_sub')
-    self.assertTrue(instr.has_unbounded_source(p))
+    self.assertTrue(instr.has_unbounded_sources(p))
 
   def test_not_has_unbounded_source(self):
     p = beam.Pipeline(interactive_runner.InteractiveRunner())
     with tempfile.NamedTemporaryFile(delete=False) as f:
       f.write(b'test')
     _ = p | 'ReadBoundedSource' >> beam.io.ReadFromText(f.name)
-    self.assertFalse(instr.has_unbounded_source(p))
+    self.assertFalse(instr.has_unbounded_sources(p))
+
+  def test_background_caching_pipeline_proto(self):
+    p = beam.Pipeline(interactive_runner.InteractiveRunner())
+
+    # Test that the two ReadFromPubSub are correctly cut out.
+    a = p | 'ReadUnboundedSourceA' >> beam.io.ReadFromPubSub(
+        subscription='projects/fake-project/subscriptions/fake_sub')
+    b = p | 'ReadUnboundedSourceB' >> beam.io.ReadFromPubSub(
+        subscription='projects/fake-project/subscriptions/fake_sub')
+
+    # Add some extra PTransform afterwards to make sure that only the unbounded
+    # sources remain.
+    c = (a, b) | beam.CoGroupByKey()
+    _ = c | beam.Map(lambda x: x)
+
+    instrumenter = instr.PipelineInstrument(p)
+    instrumenter.instrument()
+    actual_pipeline = instrumenter.background_caching_pipeline_proto()
+
+    # Now recreate the expected pipeline, which should only have the unbounded
+    # sources.
+    p = beam.Pipeline(interactive_runner.InteractiveRunner())
+    a = p | 'ReadUnboundedSourceA' >> beam.io.ReadFromPubSub(
+        subscription='projects/fake-project/subscriptions/fake_sub')
+    instrumenter._write_cache(p, a)
```

Comment:
If you want to use `_write_cache` as a static function in tests, maybe we can extract the logic into `write_cache` in the module and reuse it in `_write_cache` so that everything is self contained in `PipelineInstrument` while those logic are also available as static functions outside of `PipelineInstrument`.
An example is static `cacheable_key` in `pipeline_instrument` module and `_cacheable_key` in `PipelineInstrument` class.

Reply:
I'll choose not to use the _write_cache API then. Using it feels a little like making it too much of a change detector test, rather than actually testing the underlying code. I changed the test to use the cache_manager.WriteCache instead.